controller.go

```diff
@@ -0,0 +1,236 @@
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+		return
+	}
+
+	// Get nodes known by kubernetes cluster.
+	nodes, err := c.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		glog.Errorf("Error monitoring node status: %v", err)
+		return
+	}
+
+	for i := range nodes.Items {
+		var currentReadyCondition *v1.NodeCondition
+		node := &nodes.Items[i]
+
+		glog.Infof("Checking node %s", node.Name)
+
+		// Try to get the current node status
+		// If node status is empty, then kubelet has not posted ready status yet. In this case,
+		// try nodeStatusUpdateRetry times. If still nothing than give up this node and process next node.
+		for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+			_, currentReadyCondition = nodeutilv1.GetNodeCondition(&node.Status, v1.NodeReady)
+			if currentReadyCondition != nil {
+				break
+			}
+			name := node.Name
+			node, err = c.kubeClient.CoreV1().Nodes().Get(name, metav1.GetOptions{})
+			if err != nil {
+				glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
+				break
+			}
+			time.Sleep(retrySleepTime)
+		}
+		if currentReadyCondition == nil {
+			glog.Errorf("Update status of Node %v from Controller exceeds retry count or the Node was deleted.", node.Name)
+			continue
+		}
+		// If the known node status says that Node is NotReady, then check if the node has been removed
+		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately.
+		if currentReadyCondition != nil {
+			if currentReadyCondition.Status != v1.ConditionTrue {
+				// Check with the cloud provider to see if the node still exists. If it
+				// doesn't, delete the node immediately.
+				exists, err := ensureNodeExists(instances, node)
+				if err != nil {
+					glog.Errorf("Error getting data for node %s from cloud provider: %v", node.Name, err)
+					continue
+				}
+
+				if exists {
+					// Continue checking the remaining nodes since the current one is fine.
+					continue
+				}
+
+				glog.Infof("Deleting node since it is no longer present in cloud provider: %s", node.Name)
+
+				go func(nodeName string) {
+					defer utilruntime.HandleCrash()
+					if err := c.kubeClient.CoreV1().Nodes().Delete(nodeName, nil); err != nil {
+						glog.Errorf("unable to delete node %q: %v", nodeName, err)
+					}
+				}(node.Name)
+
+			}
+		}
+		glog.Infof("Node %s Ready state is %s.", node.Name, currentReadyCondition.Status)
+	}
+}
+
+// Checks if the instance exists in host cluster
+func ensureNodeExists(instances cloudprovider.Instances, node *v1.Node) (bool, error) {
+	_, err := instances.ExternalID(types.NodeName(node.Name))
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			return false, nil
+		}
+		return false, fmt.Errorf("ensureNodeExists: Error fetching by NodeName: %v", err)
+	}
+
+	return true, nil
+}
```

Comment:
I think the whole provider package can be replaced with single relatively simple snippet:

```
	pod, err := client.CoreV1().Pods(namespace).Get(string(name), metav1.GetOptions{})
	if apierr.IsNotFound(err) {
		return false, nil
	} else if err != nil {
		return false, err
	} else if !podutilv1.IsPodReady(pod) {
		return false, err
	}
```

Is there any further reason to add that complexity?

Reply:
Yes, it can be replaced with one function.

And yes there is a reason for that. Please read PR description. It implements [kubernetes provider interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go)

and done similar to other [external cloud providers](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#examples).

Reply:
It's probably not necessary in our case, but generally we are trying to cover similar problem, so i though it would be good to make it aligned with community approach.